src/Npgsql/ConnectorPool.cs

```diff
@@ -49,32 +49,39 @@ sealed class ConnectorPool : IDisposable
         readonly int _min;
 
         readonly NpgsqlConnector?[] _idle;
+        readonly NpgsqlConnector?[] _open; // TODO could be a 'bitmap' whether the index has a connector or not
 
         readonly ConcurrentQueue<(TaskCompletionSource<NpgsqlConnector?> TaskCompletionSource, bool IsAsync)> _waiting;
 
         [StructLayout(LayoutKind.Explicit)]
         internal struct PoolState
         {
-            [FieldOffset(0)]
-            internal short Idle;
-            [FieldOffset(2)]
-            internal short Busy;
-            [FieldOffset(4)]
-            internal int Waiting;
-            [FieldOffset(0)]
-            internal long All;
+            [FieldOffset(0)] public int Open;
+            [FieldOffset(4)] public int Idle;
 
-            internal int Total => Idle + Busy;
+            [FieldOffset(0)] public long All;
 
-            internal PoolState Copy() => new PoolState { All = Volatile.Read(ref All) };
+            // Busy can actually be read and written non atomically, it would introduce a benign race
+            // between readers of Busy and the writer(s), connector Close, when Idle is close to zero.
+            // The writer would first decrement Open then Idle to prevent readers racing and concluding Busy == _max.
+            // However with that order a race of the Idle read and decrement could happen, having readers read and
+            // conclude Idle > 0, causing readers to loop for a non existent connector until Idle is also decremented.
+            public void Deconstruct(out int open, out int idle, out int busy)
```

Comment:
Nice use of Deconstruct for atomic extraction :)

Reply:
Yes I was happy with this!